### PR TITLE
EOD tap supports multiple exchanges

### PR DESCRIPTION
**Pull request by vasilii-gainy**
Improves EOD tap to enable multiple exchanges

**Changes:**

* Collects the list of tickers from exchanges for tap streams
* Introduces incremental bulk streams for Prices and Fundamentals
* Sends tap metrics to Datadog

### CODE REVIEW

File: docker/meltano/tap-eodhistoricaldata/tap_eodhistoricaldata/streams.py [hunk 1]
```diff
@@ -149,3 +156,135 @@ def get_records(self, context: Optional[dict]) -> Iterable[Dict[str, Any]]:
...
+class AbstractBulkStream(eodhistoricaldataStream, ABC):
+
+    @cached_property
+    def partitions(self) -> List[Dict[str, Any]]:
+        state_partitions = super().partitions
+
+        partitions = []
+        exchanges = self.config.get("exchanges", []) or ["US"]
+        for exchange in exchanges:
+            if state_partitions:
```
Comment (asmisha) [inline docker/meltano/tap-eodhistoricaldata/tap_eodhistoricaldata/streams.py]: `
state_exchanges = set([p["exchange"] for p in state_partitions])
if exchange in state_exchanges: partitions.append(filtered_partitions_for_exchange[0])
`

File: docker/meltano/tap-eodhistoricaldata/tap_eodhistoricaldata/streams.py [hunk 2]
```diff
@@ -149,3 +156,135 @@ def get_records(self, context: Optional[dict]) -> Iterable[Dict[str, Any]]:
...
+                filtered_partitions_for_exchange = [p for p in state_partitions if p["exchange"] == exchange]
+                if filtered_partitions_for_exchange:
+                    partitions.append(filtered_partitions_for_exchange[0])
+                    continue
+
+            partitions.append({"exchange": exchange})
```
Comment (asmisha) [inline docker/meltano/tap-eodhistoricaldata/tap_eodhistoricaldata/streams.py]: if state is present, we'll add one exchange twice

File: docker/meltano/tap-eodhistoricaldata/tap_eodhistoricaldata/streams.py [hunk 3]
```diff
@@ -149,3 +156,135 @@ def get_records(self, context: Optional[dict]) -> Iterable[Dict[str, Any]]:
...
+            del context["date"]
+
+    def loading_dates(self, context: Optional[dict]) -> List[str]:
+        to_date = datetime.now()
+
+        if self.get_starting_replication_key_value(context) is not None:
+            from_date = datetime.strptime(self.get_starting_replication_key_value(context), "%Y-%m-%d")
+            delta_days = to_date - from_date
+        else:
+            self.logger.info("Replication key is not found in state, processing the last 3 days")
```
Comment (asmisha) [inline docker/meltano/tap-eodhistoricaldata/tap_eodhistoricaldata/streams.py]: why only last 3 days? Does it mean for new tickers we'll have latest 3 days of prices?

File: docker/meltano/tap-eodhistoricaldata/tap_eodhistoricaldata/tap.py [hunk 4]
```diff
@@ -1,35 +1,79 @@
...
+            if self._is_stream_enabled(stream_class):
+                self.logger.info("Enabling ")
+
+        return [stream_class(tap=self) for stream_class in STREAM_TYPES if self._is_stream_enabled(stream_class)]
+
+    def _is_stream_enabled(self, stream_class) -> bool:
+        if stream_class != HistoricalPrices and stream_class != IncrementalPrices:
+            return True
+
+        is_daily_prices = "prices_stream" in self.config and "daily" == self.config["prices_stream"]
```
Comment (asmisha) [inline docker/meltano/tap-eodhistoricaldata/tap_eodhistoricaldata/tap.py]: I don't get it, when will we use this setting?

File: docker/meltano/tap-eodhistoricaldata/tap_eodhistoricaldata/streams.py (hunk 2 above)
Comment (vasilii-gainy) [reply docker/meltano/tap-eodhistoricaldata/tap_eodhistoricaldata/streams.py]: no, why?

**Review by asmisha — APPROVED** [review]

**Pull request merged**